File: Dynamic Programming/Frog Jump - (i+1) (i+2)/solution_TD.cpp
#include <bits/stdc++.h> 

// Function to recursively calculate the minimum cost to reach the last stone
// with the given jumps
int solve(int n, vector<int> &heights, vector<int> &dp ){
    // Base case: If there is only one stone, the cost is 0
    if (n == 0) return 0;

    // Check if the solution for this state is already computed
    if (dp[n] != -1) return dp[n]; 

    // Calculate the cost of reaching the current stone by jumping one step
    int left = solve(n-1, heights, dp) + abs(heights[n] - heights[n-1]);
    
    // Calculate the cost of reaching the current stone by jumping two steps
    int right = INT_MAX;
    if (n-2 >= 0) right = solve(n-2, heights, dp) + abs(heights[n] - heights[n-2]);
    
    // Store the minimum cost for the current state in the DP array
    return dp[n] = min(left, right);
}

// Function to find the minimum cost to reach the last stone
int frogJump(int n, vector<int> &heights)
{
    // DP array to store the minimum cost for each stone
    vector<int> dp(n+1, -1);
    
    // Call the recursive function to find the minimum cost
    return solve(n-1, heights, dp);
}

File: Dynamic Programming/Minimum Cost of Climbing Stairs/solution_TD.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        # Calculate the minimum cost of climbing stairs using dynamic programming.

        # Get the length of the cost array.
        n = len(cost)

        # Recursive function to calculate the minimum cost of climbing stairs.
        def climb(cost, n, dp):
            # Base case: If we are below the first step, the cost is zero.
            if n < 0: 
                return 0
            # Base cases: If we are on the first or second step, return the cost of that step.
            if n == 0:
                return cost[n]
            if n == 1:
                return cost[n]

            # If the minimum cost for the current step is already calculated, return it.
            if dp[n] != -1:
                return dp[n]

            # Calculate the minimum cost for the current step by considering two options:
            # 1. Climbing from the previous step.
            # 2. Climbing from two steps back.
            dp[n] = cost[n] + min(climb(cost, n-1, dp), climb(cost, n-2, dp))

            # Return the calculated minimum cost for the current step.
            return dp[n]

        # Initialize an array to store the minimum cost for each step, initially set to -1.
        dp = [-1]*n

        # Return the minimum cost of climbing stairs by considering starting from the last or second-to-last step.
        return min(climb(cost, len(cost)-1, dp), climb(cost, len(cost)-2, dp))
